Add mutation for moving a project to a parent

mafesan opened the pull request:
PR related to issue #42.

Part of the work from this PR were splitted into smaller PRs:  #44, #45 and #46.

These changes include the following features: 
  * Move a project to a parent project.
  * When creating a new project, the parent can be set.
  * There is two operation types (for logging purposes) named `LINK` and `UNLINK`. 

Some considerations and restrictions regarding parent projects:

* If a project is not linked to any parent project, it is considered as "root" project.
* Once a project is linked to a parent project, it cannot be a root project anymore (the parent project can be changed but not removed).
* A project under a given root project (e.g. Proj1) cannot have a parent belonging to another root project (e.g. Proj2).
* A project cannot have a parent which belongs to a different ecosystem.
* A project cannot have as parent any "grandson" project (or any of its children) from the same root project.

During the first iteration we defined the possibility of moving a project to a different ecosystem. That implies: 
* All projects must reference to the ecosystem they belong to.
* If a project is moved to another ecosystem, all its children have to be updated too.

Due to the complexity of this feature, we decided to postpone it for the future.

About getting the root given a project and getting all the descendants from a given root project, I have included a couple of basic methods to do so, but the idea is to optimize these processes using raw, recursive queries.

---
Example: Link a project to a parent project:

```
mutation {
  moveProject(fromProjectId: 6, toProjectId: 3) {
    project {
      id
      name
      parentProject {
        id
        name
      }
      ecosystem {
        id
        name
      }
    }
  }
}
```

sduenas commented on bestiary/core/api.py:
```diff
@@ -157,3 +211,161 @@ def delete_ecosystem(ctx, ecosystem_id):
+    if parent_id is None:
+        raise InvalidValueError(msg="'parent_id' cannot be None")
+
+    trxl = TransactionsLog.open('link_projects', ctx)
+
+    try:
+        project = find_project(project_id)
+        parent_project = find_project(parent_id)
+    except ValueError as e:
+        raise InvalidValueError(msg=str(e))
+    except NotFoundError as exc:
+        raise exc
+
+    link_parent_project_db(trxl, project, parent_project)
+
+    trxl.close()
+
+    return project
+
+
+@django.db.transaction.atomic
+def unlink_parent_project(ctx, project_id):
+    """"""
+    if project_id is None:
+        raise InvalidValueError(msg="'project_id' cannot be None")
+
+    trxl = TransactionsLog.open('link_projects', ctx)
+
+    try:
+        project = find_project(project_id)
+    except ValueError as e:
+        raise InvalidValueError(msg=str(e))
+    except NotFoundError as exc:
+        raise exc
+
+    unlink_parent_project_db(trxl, project)
+
+    trxl.close()
+
+    return project
```
It's better to define this as `move_project`. So you don't need to have two functions doing practically the same. Also you can add here the ecosystem to do the same at once. Having the next prototype for the function:

`def move_project(ctx, from_project_id, to_project_id=None, to_ecosystem_id=None)`

we can have the next behaviour:

- When `to_project_id` and `to_ecosystem_id` are `None`, the project will be a root project.
- When `to_project_id` is not None, the project will be a child of `to_project_id`
- When `to_ecosystem_id` is not None, the project will be moved to that ecosystem (as a root or not depending if `to_project_id` was given).

mafesan replied:
Ok @sduenas, thank you for the idea!

sduenas commented on bestiary/core/models.py:
```diff
@@ -178,6 +180,42 @@ def __str__(self):
+    parent_project = ForeignKey("Project",
+                                parent_link=True,
+                                null=True,
+                                on_delete=SET_NULL)
```
Maybe it's better if this is set to `ON DELETE CASCADE`, so child projects will be deleted automatically.

mafesan replied:
I did this on purpose, as I thought maybe the user would like to remove only a "child" project but not its parent (and if the parent has a parent too, it also will be deleted). I'm changing the behavior then!

sduenas replied:
If you remove the children, you don't remove the parent. `on_delete` works when you delete the foreign key, in this case, the parent.

mafesan replied:
Oops! You are right, sorry for the confusion.

sduenas commented on bestiary/core/models.py:
```diff
@@ -202,6 +240,8 @@ class Ecosystem(EntityBase):
     description = CharField(max_length=MAX_SIZE_CHAR_FIELD,
                             null=True,
                             help_text='Brief text describing the ecosystem')
+    projects = ManyToManyField(Project,
```
I see this interesting but I'm not sure if it creates more problems and adds complexity to the system. The good thing is we can have the same project in different ecosystems, so ecosystems are synchronized. The main problem is what happens when the user wants to have almost the same project structure but not equal.

mafesan replied:
I see your point. As we began to define in #39, a project is just a set of existing end-point views, so if the user wants to have a very similar project structure, the project can be copied selecting which end-point views remain, which ones don't and which ones are new. WDYT? Is this something you want to avoid? Thanks!

sduenas replied:
Not sure. I understand the point about the views but I don't think projects structure should be a fixed one. Once you define one and you have it in two ecosystems, that means any change you make in one ecosystem is changed in the second one too.

mafesan replied:
You are right, I think this adds unnecessary complexity to the system. I changed the definition of the link between Ecosystems and Projects to 1:n.

mafesan commented:
I'm splitting the changes from this PR into smaller ones. Here is the first one for adding basic support for Projects and the "addProject" mutation: https://github.com/chaoss/grimoirelab-bestiary/pull/44.

mafesan commented:
I'm splitting the changes from this PR into smaller ones.
* Here is the first one for adding basic support for Projects and the `addProject` mutation (This one has to be reviewed first, as the rest depend on it): https://github.com/chaoss/grimoirelab-bestiary/pull/44 
* About removing projects: https://github.com/chaoss/grimoirelab-bestiary/pull/45
* About updating projects: https://github.com/chaoss/grimoirelab-bestiary/pull/46

mafesan commented on bestiary/core/db.py:
```diff
@@ -328,6 +332,79 @@ def delete_project(trxl, project):
+    return project
+
+
+def link_to_ecosystem(trxl, project, ecosystem):
```
This method was added because the `move_project` API method allowed to change the ecosystem where a project belongs. Although this method is not used yet, we will need it in the future so I decided to keep it (it has unit tests too).

sduenas replied:
If we're not going to use it, it is better to remove it. Probably you can linked to the comments of this PR (because if we rebase we lost the history).

sduenas commented on bestiary/core/api.py:
```diff
@@ -296,3 +314,74 @@ def delete_project(ctx, project_id):
+        project = _get_root(parent) if parent else project
+
+        return project
+    def _check_descendants(trxl, project, parent):
+        """Check if the parent project to be set is among its descendants"""
+
+        for subproject in project.subprojects.all():
+            if subproject == parent:
+                raise InvalidParentError(msg='A child project cannot be set as parent')
+            _check_descendants(trxl, subproject, parent)
```
I think it is better to have a function that would say if a project is the child of another project. Also, try to avoid recursion. I don't think it make sense to use it in this case where you can use BFS or DFS algorithms.

sduenas commented on bestiary/core/api.py:
```diff
@@ -296,3 +312,66 @@ def delete_project(ctx, project_id):
+    :raises NotFoundError: raised when the any of the projects or the ecosystem
+        do not exist in the registry
+    """
+    def _get_root(project):
```
Unless we use it someone else, probably we won't need it.

sduenas commented on bestiary/core/db.py:
```diff
@@ -328,6 +332,79 @@ def delete_project(trxl, project):
                        target=str(op_args['id']))
 
 
+def link_parent_project(trxl, project, parent_project):
+    """Link a project object with another with a child-parent relation.
+
+    This method sets the `parent_project` field value from `project` with the
```
Maybe all the checks you did in `move_project` would work better here. Otherwise, this will mean you can have a graph with loops and not a tree.

sduenas commented on bestiary/core/errors.py:
```diff
@@ -85,3 +86,10 @@ class ClosedTransactionError(BaseError):
     message = "%(msg)s"
+
+
+class InvalidParentError(BaseError):
+    """Exception raised when trying to set an invalid parent project to another project"""
+
+    code = ErrorCode.INVALID_PARENT_ERROR
+    message = "%(msg)s"
```
I don't think we need this error. A `ValueError` should be enough.

sduenas commented on bestiary/core/schema.py:
```diff
@@ -344,17 +345,19 @@ class Arguments:
         ecosystem_id = graphene.ID()
         name = graphene.String()
         title = graphene.String(required=False)
+        parent_project_id = graphene.ID(required=False)
```
I think `parent_id` works better because it's shorter. You should modify it in the model too.

sduenas commented on bestiary/core/schema.py:
```diff
@@ -344,17 +345,19 @@ class Arguments:
 
     @check_auth
-    def mutate(self, info, ecosystem_id, name, title=None):
+    def mutate(self, info, ecosystem_id, name, title=None, parent_project_id=None):
```
Rename `parent_project_id` to `parent_id`.

sduenas commented on tests/test_api.py:
```diff
@@ -1485,3 +1545,217 @@ def test_operations(self):
+
+        with self.assertRaisesRegex(InvalidParentError, PROJECT_INVALID_PARENT_EQUAL_ERROR):
+            api.move_project(self.ctx,
+                             self.project.id,
+                             self.project.id)
```
Is this the best example we can have? Probably a better test is to use a grandparent-parent-child relationship and try to move the parent under its child.

mafesan commented:
@sduenas Thank you for the feedback! I have addressed your changes and added some more minor fixes I found while I reviewed the changes. 

Now all is splitted in several commits which can be grouped into two main groups:
* The changes adding support for adding a parent when creating a new project.
* The changes related to the feature of moving a project to a parent.

I have also added a *changelog* file for each of these groups.

sduenas commented on bestiary/core/db.py:
```diff
@@ -328,6 +334,83 @@ def delete_project(trxl, project):
+        'parent_id': parent_project.id
+    }
+
+    if project == parent_project:
+        raise ValueError('Project cannot be its own parent')
+    if project.ecosystem != parent_project.ecosystem:
+        raise ValueError('Parent cannot belong to a different ecosystem')
+    if _is_descendant(parent_project, project):
+        raise ValueError('Parent cannot be a descendant')
```
I think we should also check that `parent_project != project`. Right now, the function is executed when there's no reason for.

mafesan replied:
Done!

sduenas replied:
Should you add this case to the documentation, I think it's not included.

mafesan commented:
@sduenas I have added the changes after your last comment, please check. Thank you!
